models/attribution.js

```diff
@@ -0,0 +1,286 @@
+  // note: Lizenzgenerator supports the case that artistHtml is empty
+  // by asking the user for input. This should be implemented here,
+  // but we don't get the necessary info in the /attribution endpoint.
+  // example image: https://en.wikipedia.org/wiki/File:Sodexo.svg
+  const html = self.artistHtml || '';
+  if (html.length === 0) {
+    return t(self.languageCode, 'anonymous');
+  }
+  return html;
+}
+
+function getAuthorAttributionHtml(self) {
+  return getAttributionHtml(self) || getArtistHtml(self);
+}
+
+function sanitizeHtml(html) {
+  const sanitizer = new HtmlSaniziter(html);
+  return sanitizer.sanitize();
+}
+
+function getPrintAttribution(self) {
+  let attribution = `${getAuthorAttributionText(self)} (${self.fileUrl})`;
+
+  if (!self.license.isInGroup('cc4')) {
+    attribution += `, „${fileNameFromIdentifier(self.fileTitle)}“`;
+  }
+  const editingAttribution = getEditingAttribution(self);
+  if (editingAttribution) {
+    attribution += `, ${editingAttribution}`;
+  }
+
+  const { url } = self.license;
+  if (url) {
+    attribution += ', ';
+    if (!editingAttribution && self.license.isInGroup('pd')) {
+      attribution += `${t(self.languageCode, 'pd-attribution-hint')}, ${t(
+        self.languageCode,
+        'check-details'
+      )} Wikimedia Commons: `;
+    }
+    attribution += url;
+  }
+
+  return attribution;
+}
+
+function getHtmlLicense(self) {
+  const { url } = self.license;
+  if (url) {
+    return `<a href="${url}" rel="license">${self.license.name}</a>`;
+  }
+
+  return self.license.name;
+}
+
+function getHtmlAttribution(self) {
+  let attributionLink;
+  let editingAttribution = getEditingAttribution(self);
+
+  if (!editingAttribution && self.license.isInGroup('pd')) {
+    attributionLink = `, ${t(self.languageCode, 'pd-attribution-hint')}`;
+    if (self.license.url) {
+      attributionLink += `, ${t(self.languageCode, 'check-details')} <a href="${
+        self.license.url
+      }" rel="license">Wikimedia Commons</a>`;
+    }
+  } else {
+    attributionLink = `, ${getHtmlLicense(self)}`;
+  }
+
+  if (editingAttribution) {
+    editingAttribution = `, ${editingAttribution}`;
+  }
+
+  return (
+    `${getAuthorAttributionHtml(self)}, ` +
+    `<a href="${self.fileUrl}">${fileNameFromIdentifier(
+      self.fileTitle
+    )}</a>${editingAttribution}${attributionLink}`
+  );
+}
+
+function getAttributionAsTextWithLinks(self) {
+  let urlSnippet = '';
+  const { url } = self.license;
+  if (url) {
+    urlSnippet = `, ${url}`;
+  }
+
+  let editingAttribution = getEditingAttribution(self);
+  if (editingAttribution) {
+    editingAttribution = `, ${editingAttribution}`;
+  }
+
+  return `${getAuthorAttributionText(self)} (${self.fileUrl}), „${fileNameFromIdentifier(
+    self.fileTitle
+  )}“${editingAttribution}${urlSnippet}`;
+}
+
+class Attribution {
+  constructor(params) {
+    validateParams(params);
+    Object.assign(this, params);
+
+    this.fileTitle = params.fileInfo.title;
```

Comment:
Can we not just move this up and then assign all things at once? E.g. like so (only rough sketch):
```javascript
validateParams(params);
const { title: fileTitle, rawUrl: fileUrl } = params.fileInfo;
// do the HTML things...
Object.assign(this, {...params, fileTitle, fileUrl, artistHtml, attributionHtml });